### ISSUE TITLE
[Documentation question] Streaming decompression example doesn't consider `output.pos == output.size` condition

### ISSUE THREAD

**Issue by yeetfield** [facebook/zstd#3654]
**Describe the bug**
The [manual chapter](http://facebook.github.io/zstd/zstd_manual.html#Chapter9) on streaming decompression describes the use of `ZSTD_decompressStream`. It says that if `output.pos == output.size`, after a call to `ZSTD_decompressStream`, then it's necessary to call `ZSTD_decompressStream` again due to the possibility of unflushed contents within zstd's internal buffers. However, the [streaming decompression example](https://github.com/facebook/zstd/blob/dev/examples/streaming_decompression.c) doesn't do that. I've also verified locally that decompression does still work, with a variety of inputs, without ever considering `output.pos == output.size`. 

Is this just a matter of updating the documentation, or am I misunderstanding the documentation?

**To Reproduce**
1. Run the [streaming example](https://github.com/facebook/zstd/blob/dev/examples/streaming_decompression.c)
2. Observe that it functions correctly

**Expected behavior**
I expect to see the streaming example consider the case where `output.pos == output.size`.

**Screenshots and charts**
N/A

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Version: 20.04
 - Compiler: g++ 12.2.0
 - Flags: can provide if needed
 - Other relevant hardware specs: N/A
 - Build system: can provide if needed

**Additional context**
N/A

**Comment by Cyan4973** [comment]
This is a relatively complex topic.

The way you are presenting the situation is focused on file streaming operation. In essence, let's consume the entire input, it will produce a file. Continue until you run out of input data to decompress, the file is regenerated.
This is what the streaming decompression example does.

But there are other possible scenarios employing the streaming interface.
Take a server, which receives messages from another server via a continuous pipeline, compressed using Zstandard in streaming mode.
At some point, the server receives a packet, which may end at a block boundary. 
In which case it will produce some output, corresponding to the decompressed block.
But, if the output buffer is entirely filled, it's conceivable (and actually very likely) that the decompressed block wasn't entirely flushed into the output buffer, due to lack of space, meaning there is still some data left in the internal buffers.
If the user doesn't check that possibility, it might actually have a truncated data block.
If the application logic then tries to act on this data, this is unlikely to end up well, exact damages depending on implementation details.

For file streaming, even if the block is entirely filled, there's no requirement to absolutely flush the entire block right now immediately, the rest of the block will simply be flushed at next iteration. 

Another possible strategy: the maximum size of a decompressed block is known beforehand, and is published in the frame header. Therefore, if only a single block is expected, it's possible to be absolutely certain that it was entirely flushed by providing a buffer which is large enough, because the block cannot be larger than the published maximum block size, making it possible to bypass the check. Note however that this strategy would fail if it's possible to receive more than one block in a single packet (which is technically possible with very high compression ratios, like a lot of zeroes). So there are caveats.

To summarize, there are multiple possible scenarios using the streaming interface.
But there is only one code documentation.
And therefore, we have to err on the cautious side, and detail what could happen in the worst case, and how to deal with it.
